### PR TITLE
fix(claude): keep entries whose usage.iterations[].model is null

### LINKED ISSUE #1710: Claude entries are dropped when usage.iterations[].model is null (Fable 5.1 since 2026-09-12)
## Summary

Since 2026-09-12, Claude Code writes Fable 5.1 assistant entries whose main-model iteration
inside `message.usage.iterations` carries an explicit `"model": null`. ccusage 20.0.20 discards
every such entry before deserialising it, so `daily`, `session`, `blocks`, and `claude daily
--instances` report no Fable usage for the day at all. There is no warning; the model is absent
from the output as if it had not been used.

## Environment

- ccusage 20.0.20 (npm, `npx ccusage@latest`), Linux x64
- Claude Code 2.1.268 and 2.1.269; the null appears on both, so it is not tied to a client
  release
- Only Fable 5.1 entries carry the null. Opus 4.7 entries written by the same client on the same
  day have no `model` key in their iteration and are counted

### Steps to reproduce

Put this single line into `<config>/projects/p/s.jsonl` and run
`CLAUDE_CONFIG_DIR=<config> ccusage daily --json --offline`:

```json
{"type":"assistant","timestamp":"2026-09-12T04:38:42.296Z","version":"2.1.268","sessionId":"s","requestId":"req_1","message":{"id":"msg_1","model":"claude-fable-5-1","role":"assistant","usage":{"input_tokens":2,"cache_creation_input_tokens":55866,"cache_read_input_tokens":0,"output_tokens":289,"iterations":[{"type":"message","model":null,"input_tokens":2,"output_tokens":289,"cache_creation_input_tokens":55866,"cache_read_input_tokens":0}]}}}
```

Expected: one row for 2026-09-12 with 56,157 tokens under `claude-fable-5-1`.

Actual: `"daily": []`.

Deleting the `"model":null` pair from the iteration, or replacing `null` with
`"claude-fable-5-1"`, makes the same line count as expected. A real 200-line session file with
44 such entries yields nothing; the same file with the null removed yields 1.7M tokens.

## Cause

`rust/adapters/claude/src/lib.rs`, in the per-line loop of the file loader:

```rust
if has_unsupported_null_field(line) {
    continue;
}
let Ok(data) = serde_json::from_slice::<UsageEntry>(line) else {
```

`has_unsupported_null_field` scans the raw bytes for every `:null` and rejects the line when the
field name in front of it is in `is_unsupported_nullable_field`, which lists `model`. The scan
has no notion of nesting, so it cannot tell `message.model` (which the TypeScript loader's
schema rejected when null, and which the list exists to mirror) from
`message.usage.iterations[].model`. The typed structs would accept the nested null:
`UsageIteration.model` is `Option<String>`, and `advisor_usages_from_line` already filters
iterations without a model. The pre-check rejects the line before either gets a chance.

The existing test `rejects_null_schema_fields_like_typescript_loader` covers
`message.model: null`; no test covers a null inside `iterations`.

## Proposed fix

Keep the top-level rejection and stop it from reaching into `iterations`. Two shapes, either of
which fixes the reported case:

1. In `has_unsupported_null_field`, skip the span of the `"iterations":[ ... ]` array before
   scanning for `:null`, or scan only up to the first `"iterations"` marker. The fields the
   list rejects are all top-level or directly under `message` and `message.usage`, so nothing
   the check is meant to catch lives inside the array.
2. Remove `model` from the byte list and reject a null `message.model` after typed
   deserialisation, where `UsageEntry.message.model` is already `Option<String>` and the loader
   can `continue` on `None` in the same place it handles `<synthetic>`.

A regression test in the same module:

```rust
#[test]
fn allows_null_model_inside_iterations() {
    assert!(!has_unsupported_null_field(
        br#"{"message":{"model":"m","usage":{"input_tokens":0,"iterations":[{"type":"message","model":null,"input_tokens":0,"output_tokens":0}]}}}"#
    ));
}
```

## Workaround

Feed ccusage a copy of the transcripts with the null filled in from the entry's own
`message.model` (or with the key deleted), for example per file:

```sh
jq -c 'if (try .message.usage.iterations catch null) then (.message.model // null) as $m
       | .message.usage.iterations |= map(if has("model") and .model == null
                                          then (if $m then .model = $m else del(.model) end)
                                          else . end)
       else . end'
```

### Expected behavior

_No response_

### Version

20.0.20

### PR DESCRIPTION
**Pull request by LYJW131**
Fixes #1710

## What

Since Claude Code 2.1.266, Fable 5.1 assistant entries carry `"model": null` in the main-model item of `message.usage.iterations`. `has_unsupported_null_field` scans the raw line for `:null` with no notion of nesting, so that nested null was treated like a null `message.model` and the whole line was skipped before deserialisation. `daily`, `session` and `blocks` silently lost every Fable 5.1 entry.

## Fix

`UsageIteration.model` is already `Option<String>`, so a null there is valid. The scanner now locates the `"iterations":[...]` byte span (skipping string contents and escapes, tracking bracket depth) and exempts `model` nulls inside it. Everything else is unchanged: a null `message.model` before or after the array and token-count nulls inside the array are still rejected. `daily.rs` shares the function, so both loaders are covered.

## Tests

- `allows_null_iteration_model_but_still_rejects_other_nulls`: null iteration model, missing model, empty array, strings containing brackets / escaped quotes, `message.model` null before and after the array, other nulls inside the array.
- `counts_entries_whose_iteration_model_is_null`: the repro line from #1710 loads as one `claude-fable-5-1` entry.

## Verification on real data (macOS, Claude Code 2.1.266)

- Repro line from the issue: stock 20.0.20 → `daily: []`; patched → 56,157 tokens under `claude-fable-5-1`.
- One real Fable 5.1 session file on its own: stock → nothing; patched → 28.2M tokens.
- Full `~/.claude`, today: stock 3.18M tokens (Fable output 5,596); patched 66.1M tokens (Fable output 331,854). Opus 5 numbers are identical on both, so nothing else changed.

Ran `cargo fmt --check`, `cargo clippy -p ccusage-adapter-claude --all-targets -- -D warnings` and `cargo test --workspace`.

---
## Summary by cubic
Fixes #1710 by keeping Claude entries whose `message.usage.iterations[].model` is null, so `daily`, `session`, and `blocks` no longer lose Fable 5.1 usage written by Claude Code 2.1.266+.

- Parses each nullable line once and shares that parse between null validation and typed deserialization, in both Claude loaders.
- Exempts only the direct `iterations` array of `message.usage` (and `data.message.message.usage` for AgentProgress); same-named arrays elsewhere never qualify.
- Rejects duplicate JSON members before typed deserialization, matching Serde's behavior on the raw line.
- Null `message.model`, token counts, and nested objects are still rejected, and malformed members no longer trip the pre-check.
- Adds tests covering the #1710 repro line, scoping edge cases, malformed members, and duplicate member rejection.

<sup>Written for commit 11c5135df199b450ac6883a6dd362a858261c0c8. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/ccusage/ccusage/pull/1719?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved processing of usage data containing supported `null` values.
  * Usage records with malformed or unsupported data are safely skipped instead of causing processing failures.
  * JSON usage data with varied formatting and escaped property names continues to be handled correctly.
  * Unsupported nested `null` values remain rejected, while unrelated unsupported values continue to be ignored safely.

### CODE REVIEW

File: rust/adapters/claude/src/lib.rs [hunk 1]
```diff
@@ -525,17 +526,59 @@ pub(crate) fn has_unsupported_null_field(line: &[u8]) -> bool {
...
     }
     false
 }
 
+/// Byte range of the `"iterations":[...]` array body, found without parsing the line: the
+/// scan skips string contents (with escapes) and tracks bracket depth until the array closes.
+/// Claude Code writes compact JSON, so the marker carries no whitespace, like `"usage":{`.
+fn iterations_span(line: &[u8]) -> Option<std::ops::Range<usize>> {
+    const MARKER: &[u8] = br#""iterations":["#;
+    let start = memmem::find(line, MARKER)? + MARKER.len();
```
[the remark this answers is not in the document]
Comment (LYJW131) [reply rust/adapters/claude/src/lib.rs]: Addressed in 69d337b: the span is now found by walking each `"usage"` object member by member (string contents and escapes skipped, JSON whitespace tolerated) and taking only a direct `iterations` array member, with regression tests for the whitespace form. Escaped key names (`"iter\u0061tions"`) are deliberately not decoded: key names are still compared as raw bytes, the same way the loader's existing `"usage":{` marker works, so this change does not widen or narrow that behaviour.

File: rust/adapters/claude/src/lib.rs [hunk 2]
```diff
@@ -525,17 +526,59 @@ pub(crate) fn has_unsupported_null_field(line: &[u8]) -> bool {
...
         offset = null_index + b":null".len();
     }
     false
 }
 
+/// Byte range of the `"iterations":[...]` array body, found without parsing the line: the
+/// scan skips string contents (with escapes) and tracks bracket depth until the array closes.
+/// Claude Code writes compact JSON, so the marker carries no whitespace, like `"usage":{`.
+fn iterations_span(line: &[u8]) -> Option<std::ops::Range<usize>> {
+    const MARKER: &[u8] = br#""iterations":["#;
```
[the remark this answers is not in the document]
Comment (LYJW131) [reply rust/adapters/claude/src/lib.rs]: Fixed in 69d337b: the scanner no longer searches the whole line for the marker. It walks each `"usage"` object's members at that object's own level and only accepts a direct `iterations` array member, so a same-named array earlier in the line (e.g. inside `toolUseResult`), one nested deeper inside `usage`, or the key as text inside a string never matches. Covered by `scopes_the_iteration_model_exemption_to_the_usage_object`.

[the remark this answers is not in the document]
Comment (LYJW131) [reply rust/adapters/claude/src/lib.rs]: Fixed in 69d337b together with the P1 above: whitespace between the key, the colon and the array is now tolerated, and there is a regression test for the spaced form.

File: rust/adapters/claude/src/lib.rs [hunk 3]
```diff
@@ -525,17 +529,140 @@ pub(crate) fn has_unsupported_null_field(line: &[u8]) -> bool {
...
+    let mut offset = 0;
+    while let Some(found) = memmem::find(&line[offset..], USAGE_KEY) {
+        let key_end = offset + found + USAGE_KEY.len();
+        offset = key_end;
+        let Some(body) = expect_after_whitespace(line, key_end, b':')
+            .and_then(|index| expect_after_whitespace(line, index, b'{'))
+        else {
+            continue;
+        };
+        if let Some(span) = iterations_in_object(line, body) {
```
[the remark this answers is not in the document]
Comment (ryoppippi) [reply rust/adapters/claude/src/lib.rs]: @cubic-dev-ai Fixed in e0e1c127. The scanner now walks the top-level message object, then its direct usage object, before locating iterations, so an earlier unrelated usage object cannot capture the exemption. Added the reported regression case. Validation: all 26 Claude adapter tests, cargo fmt --check, and clippy -D warnings pass.

File: rust/adapters/claude/src/lib.rs [hunk 4]
```diff
@@ -525,17 +529,140 @@ pub(crate) fn has_unsupported_null_field(line: &[u8]) -> bool {
...
+            b'"' => string_end(line, value + 1)?,
+            _ => {
+                let mut end = value;
+                while line
+                    .get(end)
+                    .is_some_and(|byte| !matches!(byte, b',' | b'}'))
+                {
+                    end += 1;
+                }
+                end - 1
```
[the remark this answers is not in the document]
Comment (ryoppippi) [reply rust/adapters/claude/src/lib.rs]: @cubic-dev-ai Fixed in e0e1c127. The scanner now rejects an empty primitive value before advancing, and the malformed-usage regression test covers this path. Validation: all 26 Claude adapter tests, cargo fmt --check, and clippy -D warnings pass.

**Comment by ryoppippi** [comment]
@pullfrog @coderabbitai @cubic-dev-ai Please re-review commit e0e1c127. It scopes the null-model exemption to the direct message.usage.iterations array and rejects malformed empty members. Local validation: all 26 Claude adapter tests, cargo fmt --check, and clippy -D warnings pass.

File: rust/adapters/claude/src/lib.rs [hunk 5]
```diff
@@ -510,30 +510,49 @@ fn is_valid_usage_entry(data: &UsageEntry) -> bool {
...
-                field_start -= 1;
-            }
-            if line.get(field_start) == Some(&b'"')
-                && is_unsupported_nullable_field(&line[field_start + 1..field_end])
-            {
-                return true;
+    if memmem::find(line, b"null").is_none() {
+        return false;
+    }
+    let Ok(root) = serde_json::from_slice::<serde_json::Value>(line) else {
```
[the remark this answers is not in the document]
Comment (ryoppippi) [reply rust/adapters/claude/src/lib.rs:522]: Fixed in 474874d0. Both direct and daily loaders now use deserialize_usage_line: null-free lines still deserialize directly, while null-containing lines parse once into serde_json::Value, run structural validation on that value, and then deserialize the typed entry from the same value. All 26 Claude adapter tests, cargo fmt --check, and clippy -D warnings pass.

File: rust/adapters/claude/src/lib.rs [hunk 6]
```diff
@@ -510,30 +507,68 @@ fn is_valid_usage_entry(data: &UsageEntry) -> bool {
...
+/// typed deserialization can share the same parse.
+pub(crate) fn deserialize_usage_line<T: DeserializeOwned>(line: &[u8]) -> Option<T> {
+    if memmem::find(line, b"null").is_none() {
+        return serde_json::from_slice(line).ok();
+    }
+    let root = serde_json::from_slice::<serde_json::Value>(line).ok()?;
+    if has_unsupported_null_field_in_value(&root) {
+        return None;
+    }
+    serde_json::from_value(root).ok()
```
[the remark this answers is not in the document]
Comment (ryoppippi) [reply rust/adapters/claude/src/lib.rs:539]: Fixed in 11c5135d. The null-validation path now uses a duplicate-aware JSON value visitor, preserving typed Serde rejection of duplicate members while still sharing one parsed value. Added a regression test; all 27 Claude adapter tests and Clippy pass.
